File: Parallel_IO_Bound.py
import threading

def write_and_read_file(index, N):
    filename = f"file_{index}_python.txt"

    # Write N integers to file
    with open(filename, "w") as f:
        for i in range(N):
            f.write(f"{i}\n")

    # Read file and sum values
    total = 0
    with open(filename, "r") as f:
        for line in f:
            total += int(line.strip())

    print(f"Thread {index} sum = {total}")

def main():
    num_threads = 4
    N = 2_500_000  # elements per thread
    threads = []

    for i in range(num_threads):
        t = threading.Thread(target=write_and_read_file, args=(i, N))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

if __name__ == "__main__":
    main()

File: Parallel_IO_Bound.cpp
#include <iostream>
#include <fstream>
#include <thread>
#include <vector>
#include <chrono>

void write_and_read_file(int index, size_t N) {
    std::string filename = "file_" + std::to_string(index) + "_c_plus_plus.txt";

    // Write N integers to file
    std::ofstream out(filename);
    for (size_t i = 0; i < N; ++i) {
        out << i << "\n";
    }
    out.close();

    // Read file and sum values
    std::ifstream in(filename);
    long long sum = 0;
    size_t value;
    while (in >> value) {
        sum += value;
    }
    in.close();

    std::cout << "Thread " << index << " sum = " << sum << std::endl;
}

int main() {
    const int num_threads = 4;
    const size_t N = 2'500'000; // elements per thread (total ~10M)
    std::vector<std::thread> threads;

    auto start = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < num_threads; ++i) {
        threads.emplace_back(write_and_read_file, i, N);
    }

    for (auto& t : threads) t.join();

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;

    std::cout << "C++: Parallel I/O benchmark with " << num_threads
              << " threads took " << elapsed.count() << " seconds.\n";

    return 0;
}
